parsing/g_constr.mlg

```diff
@@ -236,86 +203,81 @@ GRAMMAR EXTEND Gram
+          let fix = CAst.make ?loc:locf @@ CFix (lid,[dcl]) in
+          CAst.make ~loc @@ CLetIn( CAst.make ?loc:li @@ Name id,fix,None,c) }
+      | "let"; "cofix"; fx = cofix_decl; "in"; c = operconstr LEVEL "200" ->
+        { let {CAst.loc=locf;CAst.v=({CAst.loc=li;CAst.v=id} as lid,_,_,_ as dcl)} = fx in
+          let cofix = CAst.make ?loc:locf @@ CCoFix (lid,[dcl]) in
+          CAst.make ~loc @@ CLetIn( CAst.make ?loc:li @@ Name id,cofix,None,c) }
       | "let"; lb = ["("; l=LIST0 name SEP ","; ")" -> { l } | "()" -> { [] } ];
-	  po = return_type;
-	  ":="; c1 = operconstr LEVEL "200"; "in";
-          c2 = operconstr LEVEL "200" ->
-          { CAst.make ~loc @@ CLetTuple (lb,po,c1,c2) }
+        po = return_type; ":="; c1 = operconstr LEVEL "200"; "in";
+        c2 = operconstr LEVEL "200" ->
+        { CAst.make ~loc @@ CLetTuple (lb,po,c1,c2) }
       | "let"; "'"; p = pattern LEVEL "200"; ":="; c1 = operconstr LEVEL "200";
-          "in"; c2 = operconstr LEVEL "200" ->
-	    { CAst.make ~loc @@
-            CCases (LetPatternStyle, None,    [c1, None, None],       [CAst.make ~loc ([[p]], c2)]) }
+        "in"; c2 = operconstr LEVEL "200" ->
+        { CAst.make ~loc @@
+          CCases (LetPatternStyle, None,    [c1, None, None],       [CAst.make ~loc ([[p]], c2)]) }
       | "let"; "'"; p = pattern LEVEL "200"; ":="; c1 = operconstr LEVEL "200";
-	  rt = case_type; "in"; c2 = operconstr LEVEL "200" ->
-	    { CAst.make ~loc @@
-            CCases (LetPatternStyle, Some rt, [c1, aliasvar p, None], [CAst.make ~loc ([[p]], c2)]) }
-
+        rt = case_type; "in"; c2 = operconstr LEVEL "200" ->
+        { CAst.make ~loc @@
+          CCases (LetPatternStyle, Some rt, [c1, aliasvar p, None], [CAst.make ~loc ([[p]], c2)]) }
       | "let"; "'"; p = pattern LEVEL "200"; "in"; t = pattern LEVEL "200";
-	  ":="; c1 = operconstr LEVEL "200"; rt = case_type;
-          "in"; c2 = operconstr LEVEL "200" ->
-	    { CAst.make ~loc @@
-            CCases (LetPatternStyle, Some rt, [c1, aliasvar p, Some t], [CAst.make ~loc ([[p]], c2)]) }
-      | "if"; c=operconstr LEVEL "200"; po = return_type;
-	"then"; b1=operconstr LEVEL "200";
-        "else"; b2=operconstr LEVEL "200" ->
-          { CAst.make ~loc @@ CIf (c, po, b1, b2) }
-      | c=fix_constr -> { c } ] ]
+        ":="; c1 = operconstr LEVEL "200"; rt = case_type;
+        "in"; c2 = operconstr LEVEL "200" ->
+        { CAst.make ~loc @@
+          CCases (LetPatternStyle, Some rt, [c1, aliasvar p, Some t], [CAst.make ~loc ([[p]], c2)]) }
+      | "if"; c = operconstr LEVEL "200"; po = return_type;
+        "then"; b1 = operconstr LEVEL "200";
+        "else"; b2 = operconstr LEVEL "200" ->
+        { CAst.make ~loc @@ CIf (c, po, b1, b2) }
+      | "fix"; c = fix_decls -> { let (id,dcls) = c in CAst.make ~loc @@ CFix (id,dcls) }
+      | "cofix"; c = cofix_decls -> { let (id,dcls) = c in CAst.make ~loc @@ CCoFix (id,dcls) } ] ]
```

Comment:
This adds new syntax.  Was that intentional or a mistake?  Lines 236-243 above were the original "let fix"/"let cofix" productions.

Reply:
No, no, this is not new syntax. This is the expansion of line 303 on the left (expansion of `cofix_constr`).

Reply:
Ah, you're right.  It's odd to have both the "let fix" (taking one fix_decl) and "fix" (with multiple fix_decls) forms here.  Is this an important distinction or a historical quirk?

```
binder_constr: [
| "let" "fix" fix_decl "in" operconstr200
| "let" "cofix" cofix_decl "in" operconstr200
| "fix" fix_decls
| "cofix" cofix_decls
]
``

Reply:
Not a quirk, rather a shortening. The syntax `let fix f ... in term` can be seen as a shortcut for `let f := fix f ... in term`.